### ISSUE TITLE
PHP7 ownCloud WebDAV not working

### ISSUE THREAD

**Issue by Xenopathic** [owncloud/core#21935]
~~I've spent the past couple of hours debugging why my inception ownCloud external storage isn't working, and I think I've found the problem, but I want to verify it with someone else running PHP7.~~

Steps to reproduce:

1. Set up a WebDAV connection from ownCloud to an ownCloud running on PHP7
2. Go to the main files view
3. Notice the external storage isn't visible
4. (Optional) Use Wireshark to find that only half the request actually got through, the initial PROPFIND is unauthorized (as expected) but the second request containing authentication information never comes through.

Logs say this:

    transfer closed with outstanding read data remaining

~~I think I've tracked it down to a possible regression in how PHP7 handles cURL, which (I assume) previously performed the second request immediately, but now requires a second call to `curl_exec()` to get the authorized request to send. I threw together a little test script:~~

Looks like ownCloud WebDAV on PHP7 is broken...

```
<?php

$ch = curl_init();
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_HEADER, true);
curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC | CURLAUTH_DIGEST);
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "PROPFIND");
curl_setopt($ch, CURLOPT_URL, "http://hostname/owncloud/remote.php/webdav/");
curl_setopt($ch, CURLOPT_USERPWD, "username:password");
curl_setopt($ch, CURLOPT_NOBODY, false);
curl_setopt($ch, CURLOPT_USERAGENT, "sabre-http/4.2.1 (http://sabre.io/)");
curl_setopt($ch, CURLOPT_POSTFIELDS, '<?xml version="1.0" encoding="UTF-8"?>
<d:propfind xmlns:d="DAV:">
  <d:prop xmlns:x="http://owncloud.org/ns">
    <d:getlastmodified/>
    <d:getcontentlength/>
    <d:getcontenttype/>
    <x:permissions xmlns:x="http://owncloud.org/ns"/>
    <d:resourcetype/>
    <d:getetag/>
  </d:prop>
</d:propfind>');
curl_setopt($ch, CURLOPT_HTTPHEADER, [
	"Depth: 0",
	"Content-Type: application/xml"
]);
curl_setopt($ch, CURLOPT_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);
curl_setopt($ch, CURLOPT_REDIR_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);

echo curl_exec($ch);

var_dump(curl_errno($ch));
var_dump(curl_error($ch));

curl_close($ch);
```

~~Notice how two sections are printed, the first containing `401 Unauthorized`, the second containing the (authorized) `207 Multi-Status`. It seems that Sabre HTTP only ever does the first one, so it thinks something went wrong when it sees the 401.~~

Tested on master, Arch Linux, PHP 7.0.2, cURL 7.46.0.

cc @PVince81 @evert @icewind1991

**Comment by DeepDiver1975** [comment]
What is the behaviour like with php5.6?

**Comment by DeepDiver1975** [comment]
@evert

**Comment by Xenopathic** [comment]
Hmm, it's not what I thought - var_dump is hiding the fact that the real result does actually appear further down, and it's the same (or nearly the same at least) in both PHP7 and PHP5.4. Doesn't change the fact that DAV still doesn't work - still investigating

**Comment by Xenopathic** [comment]
Turns out the issue is on the receiving end, not the sending end (so DAV connections probably work fine, but DAV connections *to* ownCloud don't). The issue only occurs when a body is sent, without a PROPFIND XML body the request correctly completes.

**Comment by Xenopathic** [comment]
OK, I think I've nailed it down. It's bad.

Seems like with a non-empty body in the request, opening `php://input` (as Sabre HTTP does) will cause PHP to terminate before printing out any content sent to `php://output`, thus the 401 has no body content, nginx gets very confused, and the whole thing breaks. My test script works if I comment out this line in the Sabre HTTP library: https://github.com/fruux/sabre-http/blob/master/lib/Sapi.php#L42

**Comment by Xenopathic** [comment]
A new test script demonstrates the problem very nicely. Put the following in a PHP file your webserver can access:

```
<?php

$fh = fopen("php://input", "r");
$str = fread($fh, 1);
echo "pong";
```

Now run `curl -i -X 'PROPFIND' http://hostname/path/to/test.php -d 'ping'

Notice it works. Now either change the fread length to 0, or comment it out entirely: it breaks.

WTF PHP :boom:

**Comment by evert** [comment]
I think this needs to be reported at bugs.php.net...

**Comment by evert** [comment]
@Xenopathic I just followed your exact steps, but I can't reproduce your error. I'm running PHP 7.0.2 (ubuntu package) on apache 2.4.12, so it might be an nginx-only thing.

**Comment by Xenopathic** [comment]
@evert Thanks for testing, I know where to look now :smile:

**Comment by evert** [comment]
Yea this is pretty frustrating and serious. Let me know if you gather any more information. Very interested!

**Comment by Xenopathic** [comment]
This is getting weird now: it *only* happens with PHP FPM running through a Unix socket, running the same php-fpm instance through TCP sockets works just fine (both from nginx). Gah!

**Comment by Xenopathic** [comment]
Hopefully we'll get some answers soon: https://bugs.php.net/bug.php?id=71466

**Comment by Xenopathic** [comment]
@evert @DeepDiver1975 Looks like the fix will be in for the next release of PHP7.0 :smile:

@carlaschroder Just FYI, ownCloud will not work with PHP7.0.2 and below with nginx using UNIX domain sockets. Might be worth documenting it - hopefully anyone new to PHP7 will immediately jump to 7.0.3 or higher once the fix is released

**Comment by MorrisJobke** [comment]
@Xenopathic So this can be closed because it is fixed upstream?

**Comment by RealRancor** [comment]
Added the info to the known issues thread in https://forum.owncloud.org/viewtopic.php?f=36&t=31252

**Comment by carlaschroder** [comment]
> ownCloud will not work with PHP7.0.2 and below with nginx using UNIX domain sockets. Might be worth documenting it - hopefully anyone new to PHP7 will immediately jump to 7.0.3 or higher once the fix is released

PHP 7 enthusiasts need to write it up for documentation Wiki. It's not going into the manuals until there is distro support for PHP 7.

**Comment by carlaschroder** [comment]
And, when all the ownCloud core apps also support PHP 7.

**Comment by LukasReschke** [comment]
I'm missing to see this in PHP 7.0.3 as per https://github.com/php/php-src/pull/1745#issuecomment-179852597. Seems not in the release branch for that one?

**Comment by RealRancor** [comment]
Its now included in 7.0.4:

https://github.com/php/php-src/commit/7806553fd39f9c972bf2204f62e67fa45730cde9
